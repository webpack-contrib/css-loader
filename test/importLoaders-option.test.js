import path from 'path';

import postcssPresetEnv from 'postcss-preset-env';

import {
  compile,
  getCompiler,
  getErrors,
  getExecutedCode,
  getModuleSource,
  getWarnings,
} from './helpers/index';

describe('"importLoaders" option', () => {
  it('should work when not specified', async () => {
    // It is hard to test `postcss` on reuse `ast`, please look on coverage before merging
    const compiler = getCompiler(
      './nested-import/source.js',
      {},
      {
        module: {
          rules: [
            {
              test: /\.css$/i,
              rules: [
                { loader: path.resolve(__dirname, '../src') },
                {
                  loader: 'postcss-loader',
                  options: { plugins: () => [postcssPresetEnv({ stage: 0 })] },
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(
      getModuleSource('./nested-import/source.css', stats)
    ).toMatchSnapshot('module');
    expect(getExecutedCode('main.bundle.js', compiler, stats)).toMatchSnapshot(
      'result'
    );
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with a value equal to "0" (`postcss-loader` before)', async () => {
    // It is hard to test `postcss` on reuse `ast`, please look on coverage before merging
    const compiler = getCompiler(
      './nested-import/source.js',
      {},
      {
        module: {
          rules: [
            {
              test: /\.css$/i,
              use: [
                {
                  loader: path.resolve(__dirname, '../src'),
                  options: { importLoaders: 0 },
                },
                {
                  loader: 'postcss-loader',
                  options: { plugins: () => [postcssPresetEnv({ stage: 0 })] },
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(
      getModuleSource('./nested-import/source.css', stats)
    ).toMatchSnapshot('module');
    expect(getExecutedCode('main.bundle.js', compiler, stats)).toMatchSnapshot(
      'result'
    );
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with a value equal to "1" (no loaders before)', async () => {
    const compiler = getCompiler('./nested-import/source.js');
    const stats = await compile(compiler);

    expect(
      getModuleSource('./nested-import/source.css', stats)
    ).toMatchSnapshot('module');
    expect(getExecutedCode('main.bundle.js', compiler, stats)).toMatchSnapshot(
      'result'
    );
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with a value equal to "1" ("postcss-loader" before)', async () => {
    // It is hard to test `postcss` on reuse `ast`, please look on coverage before merging
    const compiler = getCompiler(
      './nested-import/source.js',
      {},
      {
        module: {
          rules: [
            {
              test: /\.css$/i,
              use: [
                {
                  loader: path.resolve(__dirname, '../src'),
                  options: { importLoaders: 1 },
                },
                {
                  loader: 'postcss-loader',
                  options: { plugins: () => [postcssPresetEnv({ stage: 0 })] },
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(
      getModuleSource('./nested-import/source.css', stats)
    ).toMatchSnapshot('module');
    expect(getExecutedCode('main.bundle.js', compiler, stats)).toMatchSnapshot(
      'result'
    );
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with a value equal to "2" ("postcss-loader" before)', async () => {
    // It is hard to test `postcss` on reuse `ast`, please look on coverage before merging
    const compiler = getCompiler(
      './nested-import/source.js',
      {},
      {
        module: {
          rules: [
            {
              test: /\.css$/i,
              use: [
                {
                  loader: path.resolve(__dirname, '../src'),
                  options: { importLoaders: 2 },
                },
                {
                  loader: 'postcss-loader',
                  options: { plugins: () => [postcssPresetEnv({ stage: 0 })] },
                },
              ],
            },
          ],
        },
      }
    );
    const stats = await compile(compiler);

    expect(
      getModuleSource('./nested-import/source.css', stats)
    ).toMatchSnapshot('module');
    expect(getExecutedCode('main.bundle.js', compiler, stats)).toMatchSnapshot(
      'result'
    );
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });
});
