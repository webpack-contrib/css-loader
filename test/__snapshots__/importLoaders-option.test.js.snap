// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"importLoaders" option should work when not specified: errors 1`] = `Array []`;

exports[`"importLoaders" option should work when not specified: module 1`] = `
"// Imports
var ___CSS_LOADER_API_IMPORT___ = require(\\"../../../src/runtime/api.js\\");
var ___CSS_LOADER_AT_RULE_IMPORT_0___ = require(\\"-!../../../src/index.js!./imported.css\\");
exports = ___CSS_LOADER_API_IMPORT___(false);
exports.i(___CSS_LOADER_AT_RULE_IMPORT_0___);
// Module
exports.push([module.id, \\".foo {\\\\n  color: red;\\\\n  color: rgba(0, 0, 255, 0.9);\\\\n}\\\\n\\", \\"\\"]);
// Exports
module.exports = exports;
"
`;

exports[`"importLoaders" option should work when not specified: result 1`] = `
Array [
  Array [
    "../../src/index.js!./nested-import/imported.css",
    ".bar {
  color: blue;
  color: rgb(0 0 100% / 90%);
}
",
    "",
  ],
  Array [
    "./nested-import/source.css",
    ".foo {
  color: red;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
]
`;

exports[`"importLoaders" option should work when not specified: warnings 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "0" (\`postcss-loader\` before): errors 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "0" (\`postcss-loader\` before): module 1`] = `
"// Imports
var ___CSS_LOADER_API_IMPORT___ = require(\\"../../../src/runtime/api.js\\");
var ___CSS_LOADER_AT_RULE_IMPORT_0___ = require(\\"-!../../../src/index.js??[ident]!./imported.css\\");
exports = ___CSS_LOADER_API_IMPORT___(false);
exports.i(___CSS_LOADER_AT_RULE_IMPORT_0___);
// Module
exports.push([module.id, \\".foo {\\\\n  color: red;\\\\n  color: rgba(0, 0, 255, 0.9);\\\\n}\\\\n\\", \\"\\"]);
// Exports
module.exports = exports;
"
`;

exports[`"importLoaders" option should work with a value equal to "0" (\`postcss-loader\` before): result 1`] = `
Array [
  Array [
    "../../src/index.js?[ident]!./nested-import/imported.css",
    ".bar {
  color: blue;
  color: rgb(0 0 100% / 90%);
}
",
    "",
  ],
  Array [
    "./nested-import/source.css",
    ".foo {
  color: red;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
]
`;

exports[`"importLoaders" option should work with a value equal to "0" (\`postcss-loader\` before): warnings 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "1" ("postcss-loader" before): errors 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "1" ("postcss-loader" before): module 1`] = `
"// Imports
var ___CSS_LOADER_API_IMPORT___ = require(\\"../../../src/runtime/api.js\\");
var ___CSS_LOADER_AT_RULE_IMPORT_0___ = require(\\"-!../../../src/index.js??[ident]!./imported.css\\");
exports = ___CSS_LOADER_API_IMPORT___(false);
exports.i(___CSS_LOADER_AT_RULE_IMPORT_0___);
// Module
exports.push([module.id, \\".foo {\\\\n  color: red;\\\\n  color: rgba(0, 0, 255, 0.9);\\\\n}\\\\n\\", \\"\\"]);
// Exports
module.exports = exports;
"
`;

exports[`"importLoaders" option should work with a value equal to "1" ("postcss-loader" before): result 1`] = `
Array [
  Array [
    "../../src/index.js?[ident]!./nested-import/imported.css",
    ".bar {
  color: blue;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
  Array [
    "./nested-import/source.css",
    ".foo {
  color: red;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
]
`;

exports[`"importLoaders" option should work with a value equal to "1" ("postcss-loader" before): warnings 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "1" (no loaders before): errors 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "1" (no loaders before): module 1`] = `
"// Imports
var ___CSS_LOADER_API_IMPORT___ = require(\\"../../../src/runtime/api.js\\");
var ___CSS_LOADER_AT_RULE_IMPORT_0___ = require(\\"-!../../../src/index.js??[ident]!./imported.css\\");
exports = ___CSS_LOADER_API_IMPORT___(false);
exports.i(___CSS_LOADER_AT_RULE_IMPORT_0___);
// Module
exports.push([module.id, \\".foo {\\\\n  color: red;\\\\n  color: rgb(0 0 100% / 90%);\\\\n}\\\\n\\", \\"\\"]);
// Exports
module.exports = exports;
"
`;

exports[`"importLoaders" option should work with a value equal to "1" (no loaders before): result 1`] = `
Array [
  Array [
    "../../src/index.js?[ident]!./nested-import/imported.css",
    ".bar {
  color: blue;
  color: rgb(0 0 100% / 90%);
}
",
    "",
  ],
  Array [
    "./nested-import/source.css",
    ".foo {
  color: red;
  color: rgb(0 0 100% / 90%);
}
",
    "",
  ],
]
`;

exports[`"importLoaders" option should work with a value equal to "1" (no loaders before): warnings 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "2" ("postcss-loader" before): errors 1`] = `Array []`;

exports[`"importLoaders" option should work with a value equal to "2" ("postcss-loader" before): module 1`] = `
"// Imports
var ___CSS_LOADER_API_IMPORT___ = require(\\"../../../src/runtime/api.js\\");
var ___CSS_LOADER_AT_RULE_IMPORT_0___ = require(\\"-!../../../src/index.js??[ident]!./imported.css\\");
exports = ___CSS_LOADER_API_IMPORT___(false);
exports.i(___CSS_LOADER_AT_RULE_IMPORT_0___);
// Module
exports.push([module.id, \\".foo {\\\\n  color: red;\\\\n  color: rgba(0, 0, 255, 0.9);\\\\n}\\\\n\\", \\"\\"]);
// Exports
module.exports = exports;
"
`;

exports[`"importLoaders" option should work with a value equal to "2" ("postcss-loader" before): result 1`] = `
Array [
  Array [
    "../../src/index.js?[ident]!./nested-import/imported.css",
    ".bar {
  color: blue;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
  Array [
    "./nested-import/source.css",
    ".foo {
  color: red;
  color: rgba(0, 0, 255, 0.9);
}
",
    "",
  ],
]
`;

exports[`"importLoaders" option should work with a value equal to "2" ("postcss-loader" before): warnings 1`] = `Array []`;
